ICE: `DefId(..) is not a body: GlobalAsm`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_mir_build/src/builder/mod.rs:695:15: DefId(0:9 ~ a[83fb]::{global_asm#0}) is not a body: GlobalAsm', 'error: internal compiler error: compiler/rustc_mir_build/src/builder/mod.rs:695:15: DefId(0:9 ~ a[83fb]::{global_asm#0}) is not a body: GlobalAsm'
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
use std::arch::global_asm;

global_asm!("
.globl call_foobar
.type call_foobar,@function
.pushsection .text.call_foobar,\"ax\",@progbits
call_foobar: jmp {}
.size call_foobar, .-call_foobar
.popsection
", sym call_foobar);

````

<details><summary><strong>original code</strong></summary>
<p>

original:
````rust
//@ build-pass
// Make sure rustc doesn't ICE on asm! for a foreign architecture.
//@ needs-asm-support
//@ only-x86_64
//@ only-linux

// regression test for #96797

use std::arch::global_asm;

#[no_mangle]
fn my_func() {}

global_asm!("
.globl call_foobar
.type call_foobar,@function
.pushsection .text.call_foobar,\"ax\",@progbits
call_foobar: jmp {}
.size call_foobar, .-call_foobar
.popsection
", sym call_foobar);

fn main() {
    unsafe {
        asm!("invalid_instruction");
        //~^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!("
            invalid_instruction
        ");
        //~^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!(r#"
            invalid_instruction
        "#);
        //~^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!("
            mov eax, eax
            invalid_instruction
            mov eax, eax
        ");
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!(r#"
            mov eax, eax
            invalid_instruction
            mov eax, eax
        "#);
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!(concat!("invalid", "_", "instruction"));
        //~^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!("movaps %xmm3, (%esi, 2)", options(att_syntax));
        //~^ WARN: scale factor without index register is ignored

        asm!(
            "invalid_instruction",
        );
        //~^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!(
            "mov eax, eax",
            "invalid_instruction",
            "mov eax, eax",
        );
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!(
            "mov eax, eax\n",
            "invalid_instruction",
            "mov eax, eax",
        );
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!(
            "mov eax, eax",
            concat!("invalid", "_", "instruction"),
            "mov eax, eax",
        );
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        asm!(
            concat!("mov eax", ", ", "eax"),
            concat!("invalid", "_", "instruction"),
            concat!("mov eax", ", ", "eax"),
        );
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction'

        // Make sure template strings get separated
        asm!(
            "invalid_instruction1",
            "invalid_instruction2",
        );
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction1'
        //~^^^ ERROR: invalid instruction mnemonic 'invalid_instruction2'

        asm!(
            concat!(
                "invalid", "_", "instruction1", "\n",
                "invalid", "_", "instruction2",
            ),
        );
        //~^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction1'
        //~^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction2'

        asm!(
            concat!(
                "invalid", "_", "instruction1", "\n",
                "invalid", "_", "instruction2",
            ),
            concat!(
                "invalid", "_", "instruction3", "\n",
                "invalid", "_", "instruction4",
            ),
        );
        //~^^^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction1'
        //~^^^^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction2'
        //~^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction3'
        //~^^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction4'

        asm!(
            concat!(
                "invalid", "_", "instruction1", "\n",
                "invalid", "_", "instruction2", "\n",
            ),
            concat!(
                "invalid", "_", "instruction3", "\n",
                "invalid", "_", "instruction4", "\n",
            ),
        );
        //~^^^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction1'
        //~^^^^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction2'
        //~^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction3'
        //~^^^^^^^^ ERROR: invalid instruction mnemonic 'invalid_instruction4'

        asm!(
            "",
            "\n",
            "invalid_instruction"
        );
        //~^^ ERROR: invalid instruction mnemonic 'invalid_instruction'
    }
}

fn main() {
    extern "Rust" {
        fn call_foobar();
    }
    unsafe { call_foobar() };
}

````
</p>
</details>

Version information
````
rustc 1.87.0-nightly (bca5f37cb 2025-02-23)
binary: rustc
commit-hash: bca5f37cbded0db8d37414bb08c4b101a5f26d36
commit-date: 2025-02-23
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/bca5f37cbded0db8d37414bb08c4b101a5f26d36/compiler/rustc_mir_build/src/builder/mod.rs#L689-L701

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0425]: cannot find value `call_foobar` in this scope
  --> /tmp/icemaker_global_tempdir.KHcxLMIG7czm/rustc_testrunner_tmpdir_reporting.2pIA33Z6GYlI/mvce.rs:10:8
   |
10 | ", sym call_foobar);
   |        ^^^^^^^^^^^ not found in this scope

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.KHcxLMIG7czm/rustc_testrunner_tmpdir_reporting.2pIA33Z6GYlI/mvce.rs:10:21
   |
10 | ", sym call_foobar);
   |                     ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.KHcxLMIG7czm/rustc_testrunner_tmpdir_reporting.2pIA33Z6GYlI/mvce.rs`

error: internal compiler error: compiler/rustc_mir_build/src/builder/mod.rs:695:15: DefId(0:4 ~ mvce[6f8a]::{global_asm#0}) is not a body: GlobalAsm
  --> /tmp/icemaker_global_tempdir.KHcxLMIG7czm/rustc_testrunner_tmpdir_reporting.2pIA33Z6GYlI/mvce.rs:3:1
   |
3  | / global_asm!("
4  | | .globl call_foobar
5  | | .type call_foobar,@function
6  | | .pushsection .text.call_foobar,\"ax\",@progbits
...  |
9  | | .popsection
10 | | ", sym call_foobar);
   | |___________________^


thread 'rustc' panicked at compiler/rustc_mir_build/src/builder/mod.rs:695:15:
Box<dyn Any>
stack backtrace:
   0:     0x7622bb37d024 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hb2766718346e8914
   1:     0x7622bba162ec - core::fmt::write::h9a1616cb06c0cb0c
   2:     0x7622bca69851 - std::io::Write::write_fmt::ha495c1f1db9a43ed
   3:     0x7622bb37ce82 - std::sys::backtrace::BacktraceLock::print::hfcfc21e91f78122e
   4:     0x7622bb37f68e - std::panicking::default_hook::{{closure}}::hdbc3741f9d3f7a84
   5:     0x7622bb37f264 - std::panicking::default_hook::hc24bc002e7741aa1
   6:     0x7622ba4de2d7 - std[aff040118bb958da]::panicking::update_hook::<alloc[67af6a9cecfd7ba8]::boxed::Box<rustc_driver_impl[301a9b8b79b48566]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7622bb37ff03 - std::panicking::rust_panic_with_hook::hcb6606e9e0388155
   8:     0x7622ba519bd1 - std[aff040118bb958da]::panicking::begin_panic::<rustc_errors[ccf3a74766137fa5]::ExplicitBug>::{closure#0}
   9:     0x7622ba50dfe6 - std[aff040118bb958da]::sys::backtrace::__rust_end_short_backtrace::<std[aff040118bb958da]::panicking::begin_panic<rustc_errors[ccf3a74766137fa5]::ExplicitBug>::{closure#0}, !>
  10:     0x7622ba50dfcb - std[aff040118bb958da]::panicking::begin_panic::<rustc_errors[ccf3a74766137fa5]::ExplicitBug>
  11:     0x7622ba5235d1 - <rustc_errors[ccf3a74766137fa5]::diagnostic::BugAbort as rustc_errors[ccf3a74766137fa5]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7622baa736dc - <rustc_errors[ccf3a74766137fa5]::DiagCtxtHandle>::span_bug::<rustc_span[2a04d5a35653b919]::span_encoding::Span, alloc[67af6a9cecfd7ba8]::string::String>
  13:     0x7622baaf485a - rustc_middle[bff44984932adf4f]::util::bug::opt_span_bug_fmt::<rustc_span[2a04d5a35653b919]::span_encoding::Span>::{closure#0}
  14:     0x7622baaddd2a - rustc_middle[bff44984932adf4f]::ty::context::tls::with_opt::<rustc_middle[bff44984932adf4f]::util::bug::opt_span_bug_fmt<rustc_span[2a04d5a35653b919]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x7622baaddb9b - rustc_middle[bff44984932adf4f]::ty::context::tls::with_context_opt::<rustc_middle[bff44984932adf4f]::ty::context::tls::with_opt<rustc_middle[bff44984932adf4f]::util::bug::opt_span_bug_fmt<rustc_span[2a04d5a35653b919]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x7622b9326397 - rustc_middle[bff44984932adf4f]::util::bug::span_bug_fmt::<rustc_span[2a04d5a35653b919]::span_encoding::Span>
  17:     0x7622bab52bea - rustc_mir_build[5775ec9c077a699a]::builder::construct_error
  18:     0x7622bc3e4747 - rustc_mir_build[5775ec9c077a699a]::builder::build_mir
  19:     0x7622bba0656c - rustc_mir_transform[235952b1afe2fde9]::mir_built
  20:     0x7622bba06525 - rustc_query_impl[5384b6942a4aabb2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5384b6942a4aabb2]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 8usize]>>
  21:     0x7622bbcfc4cc - rustc_query_system[5043f4129a32abe7]::query::plumbing::try_execute_query::<rustc_query_impl[5384b6942a4aabb2]::DynamicConfig<rustc_data_structures[27e03562bfbba71a]::vec_cache::VecCache<rustc_span[2a04d5a35653b919]::def_id::LocalDefId, rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[5043f4129a32abe7]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[5384b6942a4aabb2]::plumbing::QueryCtxt, false>
  22:     0x7622bbcfc08c - rustc_query_impl[5384b6942a4aabb2]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  23:     0x7622b94205a9 - rustc_mir_transform[235952b1afe2fde9]::mir_promoted
  24:     0x7622bbea6a12 - rustc_query_impl[5384b6942a4aabb2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5384b6942a4aabb2]::query_impl::mir_promoted::dynamic_query::{closure#2}::{closure#0}, rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 16usize]>>
  25:     0x7622bbea6ce2 - rustc_query_system[5043f4129a32abe7]::query::plumbing::try_execute_query::<rustc_query_impl[5384b6942a4aabb2]::DynamicConfig<rustc_data_structures[27e03562bfbba71a]::vec_cache::VecCache<rustc_span[2a04d5a35653b919]::def_id::LocalDefId, rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 16usize]>, rustc_query_system[5043f4129a32abe7]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[5384b6942a4aabb2]::plumbing::QueryCtxt, false>
  26:     0x7622bbea691a - rustc_query_impl[5384b6942a4aabb2]::query_impl::mir_promoted::get_query_non_incr::__rust_end_short_backtrace
  27:     0x7622bcbb73ed - rustc_query_impl[5384b6942a4aabb2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5384b6942a4aabb2]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 8usize]>>
  28:     0x7622bbcfc4cc - rustc_query_system[5043f4129a32abe7]::query::plumbing::try_execute_query::<rustc_query_impl[5384b6942a4aabb2]::DynamicConfig<rustc_data_structures[27e03562bfbba71a]::vec_cache::VecCache<rustc_span[2a04d5a35653b919]::def_id::LocalDefId, rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[5043f4129a32abe7]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[5384b6942a4aabb2]::plumbing::QueryCtxt, false>
  29:     0x7622bbcfbfcc - rustc_query_impl[5384b6942a4aabb2]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7622bbd8fcba - rustc_interface[a6a17d2a4de965ee]::passes::run_required_analyses
  31:     0x7622bca657fa - rustc_interface[a6a17d2a4de965ee]::passes::analysis
  32:     0x7622bca657d9 - rustc_query_impl[5384b6942a4aabb2]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5384b6942a4aabb2]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 0usize]>>
  33:     0x7622bcac570b - rustc_query_system[5043f4129a32abe7]::query::plumbing::try_execute_query::<rustc_query_impl[5384b6942a4aabb2]::DynamicConfig<rustc_query_system[5043f4129a32abe7]::query::caches::SingleCache<rustc_middle[bff44984932adf4f]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[5384b6942a4aabb2]::plumbing::QueryCtxt, false>
  34:     0x7622bcac53f9 - rustc_query_impl[5384b6942a4aabb2]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7622bcb4e1bc - rustc_interface[a6a17d2a4de965ee]::passes::create_and_enter_global_ctxt::<core[25dd51937c36d244]::option::Option<rustc_interface[a6a17d2a4de965ee]::queries::Linker>, rustc_driver_impl[301a9b8b79b48566]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  36:     0x7622bcb41a60 - rustc_interface[a6a17d2a4de965ee]::interface::run_compiler::<(), rustc_driver_impl[301a9b8b79b48566]::run_compiler::{closure#0}>::{closure#1}
  37:     0x7622bc97eedc - std[aff040118bb958da]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[a6a17d2a4de965ee]::util::run_in_thread_with_globals<rustc_interface[a6a17d2a4de965ee]::util::run_in_thread_pool_with_globals<rustc_interface[a6a17d2a4de965ee]::interface::run_compiler<(), rustc_driver_impl[301a9b8b79b48566]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  38:     0x7622bc97f300 - <<std[aff040118bb958da]::thread::Builder>::spawn_unchecked_<rustc_interface[a6a17d2a4de965ee]::util::run_in_thread_with_globals<rustc_interface[a6a17d2a4de965ee]::util::run_in_thread_pool_with_globals<rustc_interface[a6a17d2a4de965ee]::interface::run_compiler<(), rustc_driver_impl[301a9b8b79b48566]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[25dd51937c36d244]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  39:     0x7622bc97fceb - std::sys::pal::unix::thread::Thread::new::thread_start::h4bc285b173c402c8
  40:     0x7622b6ca370a - <unknown>
  41:     0x7622b6d27aac - <unknown>
  42:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.87.0-nightly (bca5f37cb 2025-02-23) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_built] building MIR for `{global_asm#0}`
#1 [mir_promoted] promoting constants in MIR for `{global_asm#0}`
#2 [mir_borrowck] borrow-checking `{global_asm#0}`
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0425, E0601.
For more information about an error, try `rustc --explain E0425`.

```

</p>
</details>

<!--
query stack:
#0 [mir_built] building MIR for `{global_asm#0}`
#1 [mir_promoted] promoting constants in MIR for `{global_asm#0}`
#2 [mir_borrowck] borrow-checking `{global_asm#0}`
#3 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_mir_build/src/builder/mod.rs
+++ b/compiler/rustc_mir_build/src/builder/mod.rs
@@ -612,7 +612,8 @@ fn construct_error(tcx: TyCtxt<'_>, def_id: LocalDefId, guar: ErrorGuaranteed) -
         | DefKind::AssocConst
         | DefKind::AnonConst
         | DefKind::InlineConst
-        | DefKind::Static { .. } => (vec![], tcx.type_of(def_id).instantiate_identity(), None),
+        | DefKind::Static { .. }
+        | DefKind::GlobalAsm => (vec![], tcx.type_of(def_id).instantiate_identity(), None),
         DefKind::Ctor(..) | DefKind::Fn | DefKind::AssocFn => {
             let sig = tcx.liberate_late_bound_regions(
                 def_id.to_def_id(),

--- a/compiler/rustc_mir_transform/src/lib.rs
+++ b/compiler/rustc_mir_transform/src/lib.rs
@@ -316,6 +316,10 @@ fn mir_keys(tcx: TyCtxt<'_>, (): ()) -> FxIndexSet<LocalDefId> {
     // All body-owners have MIR associated with them.
     let mut set: FxIndexSet<_> = tcx.hir_body_owners().collect();
 
+    // Remove the fake bodies for `global_asm!`, since they're not useful
+    // to be emitted (`--emit=mir`) or encoded (in metadata).
+    set.retain(|&def_id| !matches!(tcx.def_kind(def_id), DefKind::GlobalAsm));
+
     // Coroutine-closures (e.g. async closures) have an additional by-move MIR
     // body that isn't in the HIR.
     for body_owner in tcx.hir_body_owners() {

--- a/tests/codegen/target-feature-inline-closure.rs
+++ b/tests/codegen/target-feature-inline-closure.rs
@@ -12,7 +12,7 @@ use std::arch::x86_64::*;
 #[cfg(target_arch = "x86_64")]
 #[target_feature(enable = "avx")]
 fn with_avx(x: __m256) -> __m256 {
-    // CHECK: fadd
+    // CHECK: fadd <8 x float>
     let add = {
         #[inline(always)]
         |x, y| unsafe { _mm256_add_ps(x, y) }
@@ -24,14 +24,10 @@ fn with_avx(x: __m256) -> __m256 {
 #[no_mangle]
 #[cfg(target_arch = "x86_64")]
 unsafe fn without_avx(x: __m256) -> __m256 {
-    // CHECK-NOT: fadd
+    // CHECK-NOT: fadd <8 x float>
     let add = {
         #[inline(always)]
         |x, y| unsafe { _mm256_add_ps(x, y) }
     };
     add(x, x)
 }
-
-// Don't allow the above CHECK-NOT to accidentally match a commit hash in the
-// compiler version.
-// CHECK-LABEL: rustc version

--- a/tests/ui/asm/global-asm-isnt-really-a-mir-body.rs
+++ b/tests/ui/asm/global-asm-isnt-really-a-mir-body.rs
@@ -1,0 +1,26 @@
+//@ revisions: emit_mir instrument cfi
+
+// Make sure we don't try to emit MIR for it.
+//@[emit_mir] compile-flags: --emit=mir
+
+// Make sure we don't try to instrument it.
+//@[instrument] compile-flags: -Cinstrument-coverage -Zno-profiler-runtime
+//@[instrument] only-linux
+
+// Make sure we don't try to CFI encode it.
+//@[cfi] compile-flags: -Zsanitizer=cfi -Ccodegen-units=1 -Clto -Ctarget-feature=-crt-static -Clink-dead-code=true
+//@[cfi] needs-sanitizer-cfi
+//@[cfi] no-prefer-dynamic
+// FIXME(#122848) Remove only-linux once OSX CFI binaries work
+//@[cfi] only-linux
+
+//@ build-pass
+//@ needs-asm-support
+
+use std::arch::global_asm;
+
+fn foo() {}
+
+global_asm!("/* {} */", sym foo);
+
+fn main() {}

--- a/tests/ui/asm/global-asm-with-error.rs
+++ b/tests/ui/asm/global-asm-with-error.rs
@@ -1,0 +1,11 @@
+// Ensure that we don't ICE when constructing the fake MIR body for a global
+// asm when the body has errors. See #137470.
+
+//@ needs-asm-support
+
+use std::arch::global_asm;
+
+global_asm!("/* {} */", sym a);
+//~^ ERROR cannot find value `a` in this scope
+
+fn main() {}

--- a/tests/ui/asm/global-asm-with-error.stderr
+++ b/tests/ui/asm/global-asm-with-error.stderr
@@ -1,5 +1,5 @@
 error[E0425]: cannot find value `a` in this scope
-  --> $DIR/global-asm-with-error.rs:6:29
+  --> $DIR/global-asm-with-error.rs:8:29
    |
 LL | global_asm!("/* {} */", sym a);
    |                             ^ not found in this scope

--- a/tests/ui/asm/global-asm-with-error.stderr
+++ b/tests/ui/asm/global-asm-with-error.stderr
@@ -1,0 +1,9 @@
+error[E0425]: cannot find value `a` in this scope
+  --> $DIR/global-asm-with-error.rs:6:29
+   |
+LL | global_asm!("/* {} */", sym a);
+   |                             ^ not found in this scope
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0425`.

